### PR TITLE
Implement a filter list for unwanted logs

### PR DESCRIPTION
**Pull request by sylr**
This patch declares an array of filters that will discard unwanted logs.

Release note:
```
apiserver logs can be flooded by error logs comming from the http.Server lib when there
is a load balancer in front of the server doing basic HTTP probing by opening and
closing the socket to ensure the server is up.

This can lead to an unwanted and undiscardable error log being generated every 5
seconds depending on the aggressivity of the load balancer probe.
```

### CODE REVIEW

**Comment by sylr** [comment]
https://github.com/golang/go/issues/26918

File: staging/src/k8s.io/apiserver/pkg/util/logs/logs.go [hunk 1]
```diff
@@ -44,8 +45,21 @@ func AddFlags(fs *pflag.FlagSet) {
 // GlogWriter serves as a bridge between the standard log package and the glog package.
 type GlogWriter struct{}
 
+var filters = [...]*regexp.Regexp{
```
Comment (neolit123) [inline staging/src/k8s.io/apiserver/pkg/util/logs/logs.go]: please use a slice here:
`[...]*regexp.Regexp{` -> `[]*regexp.Regexp{`

**Comment by neolit123** [comment]
@sylr 
please add a release note explaining this change.

/sig api-machinery
/ok-to-test

File: staging/src/k8s.io/apiserver/pkg/util/logs/logs.go (hunk 1 above)
Comment (sylr) [reply staging/src/k8s.io/apiserver/pkg/util/logs/logs.go]: All right.

**Comment by yliaog** [comment]
this looks hacky, could you set up the probe properly to avoid the logging message in the first place?

**Comment by sylr** [comment]
@yliaog I've tried but unfortunately the Azure level 4 load balancing does not provide much tuning for the probing.

**Comment by jpbetz** [comment]
I agree this looks hacky. What alternatives have been considered? How about log levels?

**Comment by sylr** [comment]
@jpbetz I don't know if there are any other options to consider as the log come from an external library (http.Server).

I've asked them if they would discard it in there lib but they told me that could be useful for debug purposes and that if I really want to get rid of it it would be by filtering it in the GlogWriter.Write() (https://github.com/golang/go/issues/26918#issuecomment-412183170).

**Comment by jpbetz** [comment]
@sylr How about instead of logging via `glog.Info`, log via `glog.V(X).Info` and then exclude those logs by setting the `-` ("log level for V logs") command line flag.  We'd need to select an appropriate `X`, maybe run `find . -name "*.go" | xargs grep "glog.V(X)"` for X=1..10 and see if you can find an appropriate level?

**Comment by sylr** [comment]
@jpbetz The log I want to discard comes from the standard Go Library `http.Server`.

It uses the default log which is setup by :

https://github.com/kubernetes/kubernetes/blob/3e2d0e0812bca251788acae2db80bf163342348c/staging/src/k8s.io/apiserver/pkg/util/logs/logs.go#L68-L73

http.Server does not use the V(X) so I don't think your idea applies here.

**Comment by sylr** [comment]
@jpbetz But instead of totally discarding the log in the filter I implemented in Write() I could use `glog.V(9).Info()` like you suggested.

**Comment by jpbetz** [comment]
@sylr As I mentioned in https://github.com/kubernetes/kubernetes/pull/67277#issuecomment-414768421, Instead of:

```
glog.Info
```

do

```
glog.V(4).Info
```

and remove the filter from the PR.

And then when you start the kube-apiserver, provide the `-v=5` flag so that verbosity 4 and lower are excluded from the log output.

I'm not certain 4 is the best verbosity here, but it seems reasonable to me at a glance. Feel free to do some further review of existing logging to see if you believe a different level would be more appropriate.

**Comment by sylr** [comment]
@jpbetz Yeah but this would filter *all* logs that use this lib or the default log library if InitLogs() as been called.

**Comment by lavalamp** [comment]
As you expected, I'm not excited about running a regexp on every log line, and I'm also not excited about hard-coding that particular exclusion for everyone.

Have you tried cobbling something out of --vmodule? https://github.com/golang/glog/blob/master/glog.go#L64-L69

Like maybe with a filename that matches and a V of -1?

**Comment by jpbetz** [comment]
@sylr I see. Yes, the way glog is used prevents setting a verbosity level scoped how you want. @lavalamp's suggestion of using `-vmodule` sounds viable though.

**Comment by sylr** [comment]
@lavalamp @jpbetz I understand the concerns.

`-vmodule` could be an option but it would, like what @jpbetz suggested, not discard unwanted logs from external libraries but would apply a verbosity level to all the logs of this lib.

It's less intrusive than implementing V(X) in staging/src/k8s.io/apiserver/pkg/util/logs/logs.go but still to much for my taste.

**Comment by lavalamp** [comment]
You should be able to use vmodule to filter by file?

**Comment by liggitt** [comment]
> You should be able to use vmodule to filter by file?

I think you'd need to switch this to `glog.InfoDepth(..., ...)` so the call frame logged would not be this single file, and then vmodule might be able to be used to better effect

**Comment by sylr** [comment]
@liggitt I've already made a separate PR for this https://github.com/kubernetes/kubernetes/pull/67298

I'm waiting for it to be merged.

**Comment by sylr** [comment]
@lavalamp The whole http.Server implementation is in the same file https://github.com/golang/go/blob/d3c3aaa61f7598f275f30fabd3749379fe0f2720/src/net/http/server.go#L1762-L1765

Using `-vmodule` would apply the whole verbosity to the logs of this lib.

**Comment by liggitt** [comment]
> Using -vmodule would apply the verbosity to all the logs of this lib. I don't want that.

That seems to be what the http.Server offers. I agree that hard-coding an exclusion for a particular message this way at this level is not great.

**Pull request closed without merging**